File: Algorithm/Sort/SelectionSort/HeapSort.cpp
/**
 * 堆排
 * 1. 遍历，将数组先排序成大根堆heapinsert
 * 2. 将数组开始结尾数互换，heapsize--
 * 3. 重新排序成大根堆heapify
 */
#include <iostream>

using namespace std;
void Swap(int &a, int &b)
{
	int temp = a;
	a = b;
	b = temp;
}
void HeapInsert(int arr[], int index);
void Heapify(int arr[], int index, int heap_size);

void HeapSort(int arr[], int len)
{
	for (size_t i = 0; i < len; i++)
	{
		HeapInsert(arr, i);
	}
	int heap_size = len;
	Swap(arr[0], arr[--heap_size]);
	while (heap_size > 0)
	{
		Heapify(arr, 0, heap_size);
		Swap(arr[0], arr[--heap_size]);
	}
}

// 从下往上
void HeapInsert(int arr[], int index)
{
	while (arr[index] > arr[(index - 1) / 2])
	{
		Swap(arr[index], arr[(index - 1) / 2]);
		index = (index - 1) / 2;
	}
}
// 从上往下
void Heapify(int arr[], int index, int heap_size)
{
	int left = 2 * index + 1;
	while (left < heap_size)
	{
		// 下方有孩子时候，谁的值大就把下标给largest
		int largest = left + 1 < heap_size && arr[left + 1] > arr[left] ? left + 1 : left;
		// 孩子中较大值再跟父比较
		largest = arr[index] < arr[largest] ? largest : index;
		if (largest == index)
		{
			break;
		}
		Swap(arr[index], arr[largest]);
		index = largest;
		left = 2 * index + 1;
	}
}
int main()
{
	int arr[] = {1, 3, 4, 2, 5, 3, 2, 6};
	int len = sizeof(arr) / sizeof(arr[0]);
	HeapSort(arr, len);
	for (size_t i = 0; i < len; i++)
	{
		cout << arr[i] << " ";
	}
	cout << endl;
	return 0;
}

File: Algorithm/src/sort/HeapSort.java
package sort;

import java.util.Arrays;

public class HeapSort {
    public static void main(String[] args) {
        int[] arr = {9, 2, 3, 2, 1, 7, 3, 10, 2, 9};
        heapSort(arr);
        System.out.println(Arrays.toString(arr));
    }

    public static void heapSort(int[] arr) {
        // 1. 形成大根堆
        for (int i = 1; i < arr.length; i++) {
            heapInsert(arr, i);
        }
        // 2. 依次将顶部数挪到尾部，做heapify
        int heapSize = arr.length;
        for (int i = heapSize; i > 0 ; i--) {
            swap(arr, 0, --heapSize);
            heapify(arr, 0, heapSize);
        }
    }

    /**
     * 在最后面新插入一个数，向上形成大根堆过程
     *
     * @param arr   为数组
     * @param index 为新插入数的下标
     */
    public static void heapInsert(int[] arr, int index) {
        int parent = (index - 1) / 2;
        while (arr[index] > arr[parent]) {
            swap(arr, index, parent);
            index = parent;
            parent = (index - 1) / 2;
        }
    }

    /**
     * 从任意位置一个出发，向下形成大根堆过程。
     *
     * @param arr      为数组
     * @param index    为插入位置
     * @param heapSize 数组大小
     */
    public static void heapify(int[] arr, int index, int heapSize) {
        int leftChild = 2 * index + 1;
        while (leftChild < heapSize) { // 如果有左孩子
            int rightChild = 2 * index + 2;
            // 如果有右孩子并且右孩子大，swapIndex为右孩子值，否则为左孩子值
            int swapIndex = (rightChild < heapSize) && (arr[leftChild] <= arr[rightChild]) ? rightChild : leftChild;
            if (arr[index] >= arr[swapIndex]) {
                break;
            }
            swap(arr, index, swapIndex);
            index = swapIndex;
            leftChild = 2 * index + 1;
        }

    }

    public static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

}
